src/Microsoft.OData.Core/Batch/ODataBatchWriter.cs

```diff
@@ -419,6 +419,14 @@ public Task FlushAsync()
         /// </remarks>
         public abstract void OnInStreamError();
 
+        /// <inheritdoc/>
+        public virtual Task OnInStreamErrorAsync()
+        {
+            // NOTE: ODataBatchWriter class is abstract and public. This method body is intended
+            // to prevent a breaking change in subclasses that provide the implementation.
+            throw new NotImplementedException();

```

Comment:
I wonder if wrapping `OnInStreamError` in a task would be a more "expected" default implementation for `OnInStreamErrorAsync`. Is that something you considered? If so, I would be curious to know what dissuaded you from that approach.

Also does the method body prevent breaking change in subclasses that provide the implementation or in those that did not provide the implementation?

Reply:
@habbes Dealing with asynchronous wrappers over synchronous methods is actually what this "true async" initiative is all about so I don't think its a good idea to wrap `OnInStreamError`.
With regard to your second point, the method body prevents a breaking change in the event that someone had subclassed `ODataBatchWriter`. The only implementation they would have provided is for the synchronous method and we are not tinkering with that.

Reply:
@gathogojr on the first point: You make a good point. But I still wonder which default is the lesser evil. My concern is whether there cases where `OnInStreamErrorAsync` is called without the knowledge or control of the developer who implemented the abstract class. I think this is usually quite likely. If this is the case, then merely updating the library without any additional changes to their program code might lead to unexpected exceptions occurring (possibly occurring in production and undetected in development). This would be quite disruptive, and a breaking change. On the other hand, if the default implementation wraps the `OnInStreamError` method, then program will retain its current behaviour even after updating the lib. While this implementation is not necessarily ideal for async, it's less disruptive and not a breaking change. If you look at the case of the built-in Stream abstract class, `Flush` is a abstract and `FlushAsync` is a virtual method which defaults to wrapping `Flush` in a task: https://source.dot.net/#System.Private.CoreLib/Stream.cs,189

Reply:
If we stick with the default of throwing an exception, then I think we should have a way to ensure that developers updating the library in their programs (which could also be libraries used by others) are aware that they need to implement that method or to ensure that that method is not inadvertently called internally by ODL.

Reply:
@habbes I think you misunderstood. The `OnStreamInErrorAsync` is not triggered automatically, just like the current `OnStreamInError` is not triggered automatically. Just adding the method does not cause it to be triggered.
[Here](https://github.com/OData/odata.net/blob/caee76d0cafbf9b107effa30c6e12920824cf406/src/Microsoft.OData.Core/JsonLight/ODataJsonLightOutputContext.cs#L950) is how the existing method is currently triggered.
Which is why throwing a `NotImplementedException` does no harm. FYI, most of the `public virtual` methods in the `public abstract` `ODataWriter` actually throw `NotImplementedException`. If as a developer you choose to provide an implementation, you'll need to implement the methods that you intend to trigger. It's no different here.

Reply:
If a developer updates the library in their programs, the method will just sit there doing nothing. It'll not be triggered at any point.

Reply:
@gathogojr I'm a bit confused. Are those methods triggered by the developer's code or by ODL code?
In the link you provided, the `OnStreamInError` method seems to be explicitly called by ODL code, so that leaves me a bit confused by what you mean when you say "just like the current `OnStreamInError` is not triggered automatically. In the PR changes, I can't see an instance where `OnStreamInErrorAsync` is called, who is responsible for calling it and where? If it's never called directly by ODL, then what's the point of having that method on the class? I feel like I'm missing something.

Reply:
@habbes I'll rephrase. The breaking change concern stems from the scenario where a developer has implemented the `ODataBatchWriter` abstract class. Let's say they call it `DeveloperODataBatchWriter`. What I meant by saying it is not triggered automatically is that, ODL wouldn't trigger `DeveloperODataBatchWriter.OnStreamInErrorAsync` at any time. The link I shared is to ODL `internal` `ODataJsonLightOutputContext` triggering the method. What I could check is if this internal class can resolve `ODataBatchWriter` from the DI container before triggering the method because that would be the only way it'd be a breaking change.

Reply:
The other concern I have with your suggestion is that the synchronous and asynchronous approach are using 2 different writers. While they share the same output stream, we verify every step of the way whether the `OutputContext` instance if set for synchronous or asynchronous writing

Reply:
@gathogojr and I had an offline chat around this topic and he clarified why this is not a breaking change. Thought I'd share that here as well.

The ODL calls `OnInStreamError` for the internal implementations of the `OutputContext` class. This internal implementations only uses internal subclasses of `ODataBatchWriter`. At no point does it use an `ODataBatchWriter` that would be passed in by third-party code or via external-facing dependency-injection. So if the user creates their own subclass of the `ODataBatchWriter`, then we can guarantee that that version will not be called by the internal ODL methods. The only exception is if the user also creates their own `OutputContext` class, but if they do this, then they are responsible for all the logic, including error handling, whether or not to use error listeners, how to use invoke error listeners, etc. So they will cater for all that (note that a lot of the built-in classes, as well as the `IODataOutputInStreamErrorListener` interface are `internal` and not accessible to the developer).

A follow-up question I had was why it was necessary to add the `OnStreamInErrorAsync` method if it's never called by ODL. The reason it was necessary is because `ODataBatchWriter` implements `IODataOutputInStreamErrorListener` and this method has been added to that interface, so either `ODataBatchWriter` or its concrete subclasses need to implement it. So a basic implementation is added to `ODataBatchWriter` so that subclasses aren't required to.